### PR TITLE
Use execle instead of popen in tacas nss to avoid shell escape exploits

### PR DESCRIPTION
**Pull request by seiferteric**
#### Why I did it

Tacacs nss library uses popen to execute useradd and usermod commands. Popen executes using a shell (/bin/sh) which is passed the command string with "-c". This means that if untrusted user input is supplied, unexpected shell escapes can occur. In this case the username supplied can be untrusted user input when logging in via ssh or other methods when tacacs is enabled. Debian has very little limitation on usernames and as such characters such as quotes, braces, $, >, | etc are all allowed. Since the nss library is run by root, any shell escape will be ran as root.

In the current community version of tacacs nss library, the issue is mitigated by the fact that the useradd command is only ran if the user is found to exist on the tacacs server, so the bad username would have to already exists there which is unlikely. However, internally (at Dell) we had to modify this behavior to support other tacacs servers that do not allow authorization messages to verify user existence prior to a successful authentication. These servers include Cisco ISE and Aruba ClearPass. In order to support these tacacs+ servers, we have to create a temporary user immediately, which means this would be a much bigger issue. 

I also plan to supply the patch to support ISE and ClearPass and as such, I would suggest taking this patch to remediate this issue first. 

#### How I did it
Replace call to popen with fork/execl of the useradd/usermod binary directly.

#### How to verify it

Install patched version of libnss-tacplus and verify that tacacs+ user login still works as expected.

### CODE REVIEW

**Comment by seiferteric** [comment]
@lguohan Can you review? Also looks like some testbed failure that has nothging to do with my changes, can we restart the failed test above? Thanks.

**Comment by adyeung** [comment]
@a-barboza pls help review the change

**Comment by a-barboza** [comment]
> @a-barboza pls help review the change

@adyeung @seiferteric @shdasari 

I recommend merging this patch. There are some minor enhancements (replace integer constants with macros which could be done). But, I recommend merging this patch. Please let me know if anything else needs to be done from my side. thank you

**Comment by seiferteric** [comment]
Okay, the test was passing, I was just looking at the old test the first time it failed. Yes please review.

**Review by a-barboza — APPROVED** [review]

**Comment by lguohan** [comment]
@liuh-80 to review

File: src/tacacs/nss/patch/0011-Replace-popen-shell-execution-with-safer-execle.patch [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
++            }
++        } while (!WIFEXITED(wstatus) && !WIFSIGNALED(wstatus));
++        if WIFEXITED(wstatus)
++            return WEXITSTATUS(wstatus);
++        else
++            return -1;
++    // Child
++    } else if(pid == 0) {
++        execl(cmd, cmd, "-G", sec_grp, name, "-g", gid, "-c", gecos, "-d", home, "-m", "-s", shell, NULL);
++        syslog(LOG_ERR, "%s: exec of %s failed!", nssname, cmd);
```
Comment (liuh-80) [inline src/tacacs/nss/patch/0011-Replace-popen-shell-execution-with-safer-execle.patch:52]: Add errno to error log?

File: src/tacacs/nss/patch/0011-Replace-popen-shell-execution-with-safer-execle.patch [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+-                command, user->secondary_grp, name, user->gid, user->info, name, user->shell);
+-            if(debug) syslog(LOG_DEBUG, "%s", buf);
+-            fp = popen(buf, "r");
+-            if(!fp || -1 == pclose(fp)) {
+-                syslog(LOG_ERR, "%s: %s popen failed errno=%d %s",
+-                        nssname, command, errno, strerror(errno));
+-                delete_conf_line(name);
+-                return -1;
+-            }
+-            if(debug)
```
Comment (liuh-80) [inline src/tacacs/nss/patch/0011-Replace-popen-shell-execution-with-safer-execle.patch:90]: Please also keep the success log.

File: src/tacacs/nss/patch/0011-Replace-popen-shell-execution-with-safer-execle.patch (hunk 1 above)
Comment (seiferteric) [reply src/tacacs/nss/patch/0011-Replace-popen-shell-execution-with-safer-execle.patch]: Okay added

File: src/tacacs/nss/patch/0011-Replace-popen-shell-execution-with-safer-execle.patch (hunk 2 above)
Comment (seiferteric) [reply src/tacacs/nss/patch/0011-Replace-popen-shell-execution-with-safer-execle.patch:89]: Okay added

**Review by liuh-80 — APPROVED** [review]

**Review by maipbui — APPROVED** [review]

**Pull request merged**